net/grpc/gateway/examples/grpc-web-java/greeter-service/src/main/java/grpcweb/examples/greeter/GreeterClient.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package grpcweb.examples.greeter;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import grpcweb.examples.greeter.GreeterOuterClass.HelloReply;
+import grpcweb.examples.greeter.GreeterOuterClass.HelloRequest;
+import java.lang.invoke.MethodHandles;
+import java.nio.ByteBuffer;
+import java.util.logging.Logger;
+import org.apache.http.HttpVersion;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.entity.ContentType;
+
+/**
+ * A simple client that requests a greeting from the {@link GreeterService}.
+ */
+public class GreeterClient {
```

Comment:
I am confused about the purpose of this - shouldn't the GreeterClient just be a simple program setting the proto message to 'foo', call the SayHello method, and print out the response? Why all these byte level validation code here?

Reply:
I see. You need a "grpc-web" client here. And this is quite a verbose way to do one in Java.

Let's discuss how we should approach here - I think making a "grpc-web client in Java" might have sent a wrong message here. We should probably re-use the JS client portion of this helloworld example here https://github.com/grpc/grpc-web/tree/master/net/grpc/gateway/examples/helloworld as the client. Take out all the Envoy and the grpc server part, and replace with yours. But we should be able to re-use the JS client. That would fit the purpose of this repo better.

Reply:
> I am confused about the purpose of this - shouldn't the GreeterClient just be a simple program setting the proto message to 'foo', call the SayHello method, and print out the response? Why all these byte level validation code here?

You are right. We don't need the validation code.
I was just trying to show how to have a simple java client.

The README.md file will say
1. To start the server which includes proxy for grpc-web service, do the following
    mvn exec:java -Dexec.mainClass="grpcweb.examples.greeter.StartServiceAndGrpcwebProxy" test

2. and, then to test the greeter app, either use the javascript (you mentioned) or use a sample java client like this:
   mvn exec:java -Dexec.mainClass="grpcweb.examples.greeter.GreeterClient" test

3. Or, you could the following to sanity check the greeter service and client by doing this (which starts the service and client, runs the client, validates the response and then shuts everything down)
   mvn exec:java -Dexec.mainClass="grpcweb.examples.greeter.GreeterClientTest" test

Or, if it is confusing, we can completely get rid of GreeterClient.java and GreeterClientTest.java

what do you think?

Reply:
I think this is fine as-is then. We can tweak this when we are working on the main README to put the right focus on things. We can just put a note here to say this is for testing only.

Reply:
added comments to say this is just for testing purposes.